### PR TITLE
rpc: fix marshalling of type-specific tx data

### LINKED ISSUE #585: Return claim-specific data in getrawtransaction RPC
C# node gives a little more information about claim transactions:
```
{
   "jsonrpc" : "2.0",
   "id" : 5,
   "result" : {
      "claims" : [
         {
            "txid" : "0x8605957cda9f86774bed580f95fc0977b00f11b1fa3cd75963efdcd6af57d440",
            "vout" : 0
         },
         {
            "vout" : 0,
            "txid" : "0x39f699a24efc2a481804a7ed85c1df1b6173a65fb6b63edfb5888cbde2c98148"
         }
      ],
      "blocktime" : 1476834933,
      "attributes" : [],
      "version" : 0,
      "blockhash" : "0x515df87354455f42f7cc2991270b09b626973ae90899056fb23bd7bdb95449fe",
      "vout" : [
         {
            "n" : 0,
            "value" : "0.62216072",
            "asset" : "0x602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7",
            "address" : "AHVgE29tkV63KMG9gccMpLRNa7FFBNDYoW"
         }
      ],
      "net_fee" : "0",
      "size" : 237,
      "confirmations" : 4819737,
      "scripts" : [
         {
            "verification" : "21029522d81a487492d3c98ce9267e8960ec4bde8f1e5f0a31ae382867b39808bbf9ac",
            "invocation" : "40e8a35e4a7cadd53f6f65515528cf0851af1a9a3ecebc4675535a139c4fa846b7cc462de31432eb7c587e420cb598242ce23a567e959276045cb2839606730cea"
         }
      ],
      "type" : "ClaimTransaction",
      "txid" : "0xed5fc511eb946a181625d6e98f535e215c3b95c7406e74bdf726f7f3c709a571",
      "vin" : [],
      "sys_fee" : "0"
   }
}
```
neo-go:
```
{
   "jsonrpc" : "2.0",
   "id" : 5,
   "result" : {
      "attributes" : [],
      "vout" : [
         {
            "n" : 0,
            "address" : "AHVgE29tkV63KMG9gccMpLRNa7FFBNDYoW",
            "value" : "0.62216072",
            "asset" : "0x602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7"
         }
      ],
      "size" : 237,
      "type" : "ClaimTransaction",
      "scripts" : [
         {
            "verification" : "21029522d81a487492d3c98ce9267e8960ec4bde8f1e5f0a31ae382867b39808bbf9ac",
            "invocation" : "40e8a35e4a7cadd53f6f65515528cf0851af1a9a3ecebc4675535a139c4fa846b7cc462de31432eb7c587e420cb598242ce23a567e959276045cb2839606730cea"
         }
      ],
      "version" : 0,
      "blocktime" : 1476834933,
      "txid" : "0xed5fc511eb946a181625d6e98f535e215c3b95c7406e74bdf726f7f3c709a571",
      "confirmations" : 8548,
      "sys_fee" : "0",
      "blockhash" : "0x515df87354455f42f7cc2991270b09b626973ae90899056fb23bd7bdb95449fe",
      "vin" : [],
      "net_fee" : "0"
   }
}
```
Other transaction types should also be checked for type-specific data (file a separate issue if there is anything missing).

### PR DESCRIPTION
**Pull request by AnnaShaleva**
closes #585

### CODE REVIEW

File: pkg/core/transaction/transaction.go [hunk 1]
```diff
@@ -233,3 +238,194 @@ func (t *Transaction) Bytes() []byte {
...
+		t.Data = &StateTX{
+			Descriptors: tx.Descriptors,
+		}
+	case ContractType:
+		t.Data = &ContractTX{}
+	case IssueType:
+		t.Data = &IssueTX{}
+	}
+
+	err := t.createHash()
```
Comment (roman-khimov) [inline pkg/core/transaction/transaction.go]: `t.Hash()` will do that automatically, so this code is not very useful.

Comment (roman-khimov) [reply pkg/core/transaction/transaction.go]: Unless of course you're doing some kind of struct reuse and expecting `hash` to be already set.

File: pkg/core/transaction/transaction.go [hunk 2]
```diff
@@ -233,3 +238,194 @@ func (t *Transaction) Bytes() []byte {
...
+	}
+	if t.Hash() != tx.TxID {
+		return errors.New("can not unmarshal transaction: bad transaction id")
+	}
+
+	return nil
+}
+
+// contract is a wrapper for PublishTransaction
+type contract struct {
```
Comment (roman-khimov) [inline pkg/core/transaction/transaction.go]: `publishedContract`? `contract` is way too generic and can be confused with `state.Contract`. And maybe move its definition to `publish.go`?

File: pkg/core/transaction/transaction.go [hunk 3]
```diff
@@ -233,3 +238,194 @@ func (t *Transaction) Bytes() []byte {
...
+// code is a wrapper for PublishTransaction Code
+type code struct {
+	Hash       util.Uint160              `json:"hash,omitempty"`
+	Script     string                    `json:"script,omitempty"`
+	ParamList  []smartcontract.ParamType `json:"parameters,omitempty"`
+	ReturnType smartcontract.ParamType   `json:"returntype,omitempty"`
+}
+
+// asset is a wrapper for RegisterTransaction
+type asset struct {
```
Comment (roman-khimov) [inline pkg/core/transaction/transaction.go]: `registeredAsset` and move to `register.go`?

File: pkg/core/transaction/transaction.go [hunk 4]
```diff
@@ -233,3 +238,194 @@ func (t *Transaction) Bytes() []byte {
...
+	NeedStorage bool   `json:"needstorage,omitempty"`
+	Name        string `json:"name,omitempty"`
+	CodeVersion string `json:"version,omitempty"`
+	Author      string `json:"author,omitempty"`
+	Email       string `json:"email,omitempty"`
+	Description string `json:"description,omitempty"`
+}
+
+// code is a wrapper for PublishTransaction Code
+type code struct {
```
Comment (roman-khimov) [inline pkg/core/transaction/transaction.go]: `publishedCode`?

File: pkg/rpc/client/rpc_test.go [hunk 5]
```diff
@@ -155,7 +156,11 @@ var rpcClientTestCases = map[string][]rpcClientTestCase{
...
-			result: func(c *Client) interface{} {
+			result:         func(c *Client) interface{} { return &result.Block{} },
```
Comment (roman-khimov) [inline pkg/rpc/client/rpc_test.go]: gofmt?

File: pkg/rpc/client/rpc_test.go [hunk 6]
```diff
@@ -248,7 +242,11 @@ var rpcClientTestCases = map[string][]rpcClientTestCase{
...
+			check: func(t *testing.T, c *Client, res interface{}) {
```
Comment (roman-khimov) [inline pkg/rpc/client/rpc_test.go]: Why did you add these `check` functions?

File: pkg/rpc/response/result/block.go [hunk 7]
```diff
@@ -15,19 +17,9 @@ type (
...
-		Attributes []transaction.Attribute `json:"attributes"`
-		VIn        []transaction.Input     `json:"vin"`
-		VOut       []transaction.Output    `json:"vout"`
-		Scripts    []transaction.Witness   `json:"scripts"`
-
-		SysFee util.Fixed8 `json:"sys_fee"`
-		NetFee util.Fixed8 `json:"net_fee"`
-
-		Nonce uint32 `json:"nonce,omitempty"`
+		Transaction *transaction.Transaction
```
Comment (roman-khimov) [inline pkg/rpc/response/result/block.go]: Not sure it's a good change, we're missing original `TxID` and `Size` field, although it's possible to get them via other means. @im-kulikov: are you using any of these fields directly?

File: pkg/rpc/response/result/block.go [hunk 8]
```diff
@@ -77,32 +69,72 @@ func NewBlock(b *block.Block, chain core.Blockchainer) Block {
...
+		Scripts:    t.Transaction.Scripts,
+		Trimmed:    t.Transaction.Trimmed,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// We have to keep both transaction.Transaction and tx at the same level in json in order to match C# API,
+	// so there's no way to marshall Tx correctly with standard json.Marshaller tool.
+	closingBraceIndex := bytes.LastIndexByte(output, '}')
```
Comment (roman-khimov) [inline pkg/rpc/response/result/block.go]: Normally it's `len(output) - 1`, so I'd probably do something like
```
if output[len(output) - 1] != '}' || txBytes[0] != '{' {
    return nil, errors.New("can't merge internal jsons")
}
output[len(output)-1] = ','
output = append(output, txBytes[1:]...)
```

File: pkg/rpc/response/result/tx_raw_output.go [hunk 9]
```diff
@@ -1,24 +1,24 @@
...
 type TransactionOutputRaw struct {
-	*transaction.Transaction
-	TxHash        util.Uint256 `json:"txid"`
-	Size          int          `json:"size"`
-	SysFee        util.Fixed8  `json:"sys_fee"`
-	NetFee        util.Fixed8  `json:"net_fee"`
-	Blockhash     util.Uint256 `json:"blockhash,omitempty"`
-	Confirmations int          `json:"confirmations,omitempty"`
-	Timestamp     uint32       `json:"blocktime,omitempty"`
+	Transaction   *transaction.Transaction
```
Comment (roman-khimov) [inline pkg/rpc/response/result/tx_raw_output.go]: Can we keep it embedded?

File: pkg/rpc/response/result/tx_raw_output.go [hunk 10]
```diff
@@ -31,12 +31,78 @@ func NewTransactionOutputRaw(tx *transaction.Transaction, header *block.Header,
...
 		SysFee:        chain.SystemFee(tx),
 		NetFee:        chain.NetworkFee(tx),
 		Blockhash:     header.Hash(),
 		Confirmations: confirmations,
 		Timestamp:     header.Timestamp,
 	}
 }
+
+// transactionOutputRaw is an auxilliary struct for proper TransactionOutputRaw marshaling.
+type transactionOutputRaw struct {
```
Comment (roman-khimov) [inline pkg/rpc/response/result/tx_raw_output.go]: `TransactionMetadata`? And then maybe

```
type TransactionOutputRaw struct {
 *transaction.Transaction
 TransactionMetadata
}
```
?

File: pkg/rpc/response/result/tx_raw_output.go [hunk 11]
```diff
@@ -31,12 +31,78 @@ func NewTransactionOutputRaw(tx *transaction.Transaction, header *block.Header,
...
+		SysFee:        t.SysFee,
+		NetFee:        t.NetFee,
+		Blockhash:     t.Blockhash,
+		Confirmations: t.Confirmations,
+		Timestamp:     t.Timestamp,
+	})
+	if err != nil {
+		return nil, err
+	}
+	txBytes, err := json.Marshal(&transaction.Transaction{
```
Comment (roman-khimov) [inline pkg/rpc/response/result/tx_raw_output.go]: `json.Marshal(t.Transaction)`?

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/rpc/client/rpc_test.go (hunk 6 above)
Comment (AnnaShaleva) [reply pkg/rpc/client/rpc_test.go]: I wasn't aware of the fact that we can automatically fill internal fields `hash` and `verificationHash` using `trandaction.Hash()` method, so the difference between expected and actual transactions was in these two fields. Now fixed.

File: pkg/rpc/response/result/block.go [hunk 12]
```diff
@@ -48,7 +45,7 @@ type (
 
 		Script transaction.Witness `json:"script"`
 
-		Tx []Tx `json:"tx"`
+		Tx []*Tx `json:"tx"`
```
Comment (roman-khimov) [inline pkg/rpc/response/result/block.go]: `Tx` is quite simple, do we need a pointer here for some reason?

Comment (AnnaShaleva) [reply pkg/rpc/response/result/block.go]: No, we don't. Fixed

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**